src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs

```diff
@@ -0,0 +1,560 @@
+        {
+            internal static Task Register(IPoll.Pollable pollable)
+            {
+                var handle = pollable.Handle;
+                pollable.Handle = 0;
+                return CallRegister((Thread)null!, handle);
+
+                [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "Register")]
+                static extern Task CallRegister(Thread t, int handle);
+            }
+        }
+
+        private sealed class InputStream : Stream
+        {
+            private ITypes.IncomingBody body;
+            private IStreams.InputStream stream;
+            private int offset;
+            private byte[]? buffer;
+            private bool closed;
+
+            public InputStream(ITypes.IncomingBody body)
+            {
+                this.body = body;
+                this.stream = body.Stream();
+            }
+
+            ~InputStream()
+            {
+                Dispose(false);
+            }
+
+            public override bool CanRead => true;
+            public override bool CanWrite => false;
+            public override bool CanSeek => false;
+            public override long Length => throw new NotImplementedException();
+            public override long Position
+            {
+                get => throw new NotImplementedException();
+                set => throw new NotImplementedException();
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                stream.Dispose();
+                ITypes.IncomingBody.Finish(body);
+            }
+
+            public override long Seek(long offset, SeekOrigin origin)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override void Flush()
+            {
+                // ignore
+            }
+
+            public override void SetLength(long length)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override int Read(byte[] buffer, int offset, int length)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override void Write(byte[] buffer, int offset, int length)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override async Task<int> ReadAsync(
+                byte[] bytes,
+                int offset,
+                int length,
+                CancellationToken cancellationToken
+            )
+            {
+                // TODO: handle `cancellationToken`
+                while (true)
+                {
+                    if (closed)
+                    {
+                        return 0;
+                    }
+                    else if (this.buffer == null)
+                    {
+                        try
+                        {
+                            // TODO: should we add a special case to the bindings generator
```

Comment:
JavaScript has nice "Bring Your Own Buffer" idea for working with streams.
https://developer.mozilla.org/en-US/docs/Web/API/ReadableStreamBYOBReader

Reply:
Thanks for the link.  The plan for WASIp3 is that the Component Model will have a built-in stream type that allows readers to supply their own buffer (rather than having the callee allocate a buffer and return it to the caller), which we'll be able to generate efficient bindings for now.  I expect that for WASIp2, there will be at least one copy involved, although there are still some fancy tricks we can do in the bindings generator (e.g. hook into the `cabi_realloc` function and make it return an already-allocated buffer in certain cases).

For now though, I'm going to focus on just making everything work; we can always optimize later.